compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp

```diff
@@ -174,46 +178,61 @@ class VulkanSPIRVTargetBackend : public TargetBackend {
 
     ModuleOp innerModuleOp = variantOp.getInnerModule();
     auto spirvModuleOps = innerModuleOp.getOps<spirv::ModuleOp>();
-    if (!llvm::hasSingleElement(spirvModuleOps)) {
-      return variantOp.emitError()
-             << "should only contain exactly one spirv.module op";
-    }
-    auto spvModuleOp = *spirvModuleOps.begin();
-    if (!options.dumpIntermediatesPath.empty()) {
-      std::string assembly;
-      llvm::raw_string_ostream os(assembly);
-      spvModuleOp.print(os, OpPrintingFlags().useLocalScope());
-      dumpDataToPath(options.dumpIntermediatesPath, options.dumpBaseName,
-                     variantOp.getName(), ".mlir", assembly);
+    if (spirvModuleOps.empty()) {
+      return variantOp.emitError() << "should contain some spirv.module ops";
     }
 
     FlatbufferBuilder builder;
     iree_hal_spirv_ExecutableDef_start_as_root(builder);
 
-    // Serialize the spirv::ModuleOp into the binary that we will embed in the
-    // final FlatBuffer.
-    SmallVector<uint32_t, 256> spvBinary;
-    if (failed(spirv::serialize(spvModuleOp, spvBinary)) || spvBinary.empty()) {
-      return variantOp.emitError() << "failed to serialize spirv.module";
-    }
-    if (!options.dumpBinariesPath.empty()) {
-      dumpDataToPath<uint32_t>(options.dumpBinariesPath, options.dumpBaseName,
-                               variantOp.getName(), ".spv", spvBinary);
-    }
-
-    auto spvCodeRef = flatbuffers_uint32_vec_create(builder, spvBinary.data(),
-                                                    spvBinary.size());
-
-    // The runtime uses ordinals instead of names. We provide the list of entry
-    // point names here that are then passed in VkShaderModuleCreateInfo.
     SmallVector<StringRef> entryPointNames;
     SmallVector<uint32_t> subgroupSizes;
+    SmallVector<iree_hal_spirv_ShaderModuleDef_ref_t> shaderModuleRefs;
+    SmallVector<uint32_t> shaderModuleIndices;
     SmallVector<iree_hal_spirv_FileLineLocDef_ref_t> sourceLocationRefs;
     bool hasAnySubgroupSizes = false;
-    spvModuleOp.walk([&](spirv::EntryPointOp exportOp) {
-      entryPointNames.push_back(exportOp.getFn());
 
-      auto fn = spvModuleOp.lookupSymbol<spirv::FuncOp>(exportOp.getFn());
+    // Iterate over all spirv.module ops and encode them into the FlatBuffer
+    // data structure.
+    for (spirv::ModuleOp spvModuleOp : spirvModuleOps) {
+      // Currently the spirv.module op should only have one entry point. Get it.
+      auto spirvEntryPoints = spvModuleOp.getOps<spirv::EntryPointOp>();
+      if (!llvm::hasSingleElement(spirvEntryPoints)) {
+        return spvModuleOp.emitError()
+               << "expected to contain exactly one entry point";
+      }
+      spirv::EntryPointOp spvEntryPoint = *spirvEntryPoints.begin();
```

Comment:
Could user-provided spirv modules have multiple entry points? Or does linking produce multiple entry points? From the PR description, it seems like this creates separate spirv blobs each with only one entry point?

(Trying to see if this will fail to compose or be reentrant in any configurations...)

Reply:
This commit changes the underlying SPIR-V flatbuffer schema to be flexible to support multiple entry points and multiple shader module blobs in the same exectuable. We have a list to specify from which shader module each entry point comes. Therefore--
* Users can specify multiple entry points for the same shader module they provide as a blob to include. That is handled via `serializeExternalExecutable()`.
* For the normal CodeGen path though right now we are guaranteed to have one entry point per shader module. That can be improved later.